File: backend/src/services/participation.service.js
import Participation from "../models/Participation.js";
import Activity from "../models/Activity.js";
import { AppError } from "../utils/AppError.js";

export const participationService = {
  async create(userId, activityId) {
    const existing = await Participation.findByActivityAndUser(activityId, userId);
    if (existing) throw new AppError("Você já está inscrito nesta atividade", 409);
    return Participation.create({ activityId, userId, status: "pending", workloadHours: 0 });
  },

  async validatePresence(participationId, validatorId, status) {
    const participation = await Participation.findById(participationId);
    if (!participation) throw new AppError("Participação não encontrada", 404);

    let workloadHours = participation.workloadHours;
    if (status === "present") {
      const activity = await Activity.findById(participation.activity);
      if (!activity || !activity.workloadHours) {
        throw new AppError("Atividade sem carga horária definida", 400);
      }
      workloadHours = activity.workloadHours;
    } else {
      workloadHours = 0;
    }

    return Participation.updateStatus(participationId, {
      status,
      validatedBy: validatorId,
      workloadHours,
    });
  },

  listMine(userId) {
    return Participation.findByUserWithActivity(userId);
  },

  listByActivity(activityId) {
    return Participation.findByActivityWithUser(activityId);
  },
};

export default participationService;

File: backend/app/services/participation_service.py
"""Regras de participação — porta do participation.service.js."""

from __future__ import annotations

import uuid
from typing import Any

from fastapi import status
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.orm import selectinload

from app.core.errors import AppError
from app.db.models import Activity, Participation
from app.utils import serialize


async def create(
    db: AsyncSession, user_id: uuid.UUID, activity_id: uuid.UUID
) -> dict[str, Any]:
    existing = await db.scalar(
        select(Participation).where(
            Participation.activity_id == activity_id,
            Participation.user_id == user_id,
        )
    )
    if existing is not None:
        raise AppError("Você já está inscrito nesta atividade", status.HTTP_409_CONFLICT)

    participation = Participation(
        activity_id=activity_id, user_id=user_id, status="pending", workload_hours=0
    )
    db.add(participation)
    await db.commit()
    await db.refresh(participation)
    return serialize.participation(participation)


async def validate_presence(
    db: AsyncSession,
    participation_id: uuid.UUID,
    validator_id: uuid.UUID,
    new_status: str,
) -> dict[str, Any]:
    participation = await db.get(Participation, participation_id)
    if participation is None:
        raise AppError("Participação não encontrada", status.HTTP_404_NOT_FOUND)

    if new_status == "present":
        activity = await db.get(Activity, participation.activity_id)
        if activity is None or not activity.workload_hours:
            raise AppError(
                "Atividade sem carga horária definida", status.HTTP_400_BAD_REQUEST
            )
        participation.workload_hours = activity.workload_hours
    else:
        participation.workload_hours = 0

    participation.status = new_status
    participation.validated_by = validator_id
    await db.commit()
    await db.refresh(participation)
    return serialize.participation(participation)


async def list_mine(db: AsyncSession, user_id: uuid.UUID) -> list[dict[str, Any]]:
    result = await db.scalars(
        select(Participation)
        .where(Participation.user_id == user_id)
        .options(selectinload(Participation.activity))
        .order_by(Participation.created_at.desc())
    )
    return [serialize.participation_with_activity(p) for p in result]


async def list_by_activity(
    db: AsyncSession, activity_id: uuid.UUID
) -> list[dict[str, Any]]:
    result = await db.scalars(
        select(Participation)
        .where(Participation.activity_id == activity_id)
        .options(selectinload(Participation.user))
        .order_by(Participation.created_at.asc())
    )
    return [serialize.participation_with_user(p) for p in result]
